### DISCUSSION TITLE
Verifying stream with detached signature fails to validate

### DISCUSSION THREAD

**Discussion by willfarrell** [openpgpjs/openpgpjs#1271] [Q&A] [answered]
I'm trying to pass in a `Message` with a ReadableStream into `verify(...)` and even with the `readToEnd` being called afterward, `valid` remains `undefined`. I noticed there were no tests for this case. Not sure if this is a bug in the code, I'm misunderstanding the documentation or this is just unsupported. Any guidance would be greatly appreciated.

I wrote 2 tests that can be added to `/test/general/streaming.js` to duplicate this issue:
```js
// Based off `Detached sign small message` test (line:639)
it('Detached sign/verify small string stream message', async function() {
    dataArrived(); // Do not wait until data arrived.
    const dataSign = new ReadableStream({
      async start(controller) {
        controller.enqueue(util.strToUint8Array('hello '));
        controller.enqueue(util.strToUint8Array('world'));
        controller.close();
      }
    });
    const dataVerify = openpgp.stream.clone(dataSign)
    const signed = await openpgp.sign({
      message: openpgp.Message.fromBinary(dataSign),
      privateKeys: privKey,
      detached: true,
      streaming: expectedType
    });
    expect(openpgp.stream.isStream(signed)).to.equal(expectedType);
    const armoredSignature = await openpgp.stream.readToEnd(signed);
    const signature = await openpgp.readSignature({ armoredSignature });
    const verified = await openpgp.verify({
      signature,
      publicKeys: pubKey,
      message: openpgp.Message.fromBinary(dataVerify)
    });
    const dataString = await openpgp.stream.readToEnd(verified.data);
    expect(dataString).to.equal('hello world');
    expect(verified.signatures).to.exist.and.have.length(1);
    expect(await verified.signatures[0].verified).to.be.true;
  });
  
  it('Detached sign/verify small binary stream message', async function() {
    dataArrived(); // Do not wait until data arrived.
    const dataSign = new ReadableStream({
      async start(controller) {
        controller.enqueue(new Uint8Array([0x01, 0x02, 0x03]));
        controller.close();
      }
    });
    const dataVerify = openpgp.stream.clone(dataSign)
    const signed = await openpgp.sign({
      message: openpgp.Message.fromBinary(dataSign),
      privateKeys: privKey,
      detached: true,
      streaming: expectedType
    });
    expect(openpgp.stream.isStream(signed)).to.equal(expectedType);
    const armoredSignature = await openpgp.stream.readToEnd(signed);
    const signature = await openpgp.readSignature({ armoredSignature });
    const verified = await openpgp.verify({
      signature,
      publicKeys: pubKey,
      message: openpgp.Message.fromBinary(dataVerify),
      format: 'binary'
    });
    console.log(verified.signatures)
    const dataString = await openpgp.stream.readToEnd(verified.data);
    console.log(verified.signatures)
    expect(dataString).to.deep.equal(new Uint8Array([0x01, 0x02, 0x03]));
    expect(verified.signatures).to.exist.and.have.length(1);
    expect(await verified.signatures[0].verified).to.be.true;
  });
```

**Update 1:** Add in use of `stream.close` to tee data.
**Update 2:** Use `await verified.signatures[0].verified` to create valid.

**Comment by twiss** [comment, marked as answer]
Hi :wave:

A stream's contents can only be read once. If you want it to be able to be read again, you can use `openpgp.stream.clone` (which keeps an extra copy of the data in memory).

**Reply by willfarrell** [reply]
Thanks, that makes so much sense to tee the stream. I've updated the tests above with your suggestion which allows `dataString` to return the expected value. However, I'm still finding that `valid` is still undefined. Any ideas?

**Reply by twiss** [reply]
Try `await verified.signatures[0].verified` - when streaming, `valid` is not available, as we don't synchronously know the validity.

**Reply by willfarrell** [reply]
Yep, that did it. Thank you. Interesting that this needs an extra step compared to when streaming an armoredMessage. Would you like me to update the `Create and verify *detached* signatures` example to use a stream and show this? Currently, it shows using Cleartext which throws an error in v5.

**Reply by twiss** [reply]
If you want, you could instead update the "Streaming sign and verify _Uint8Array_ data" example - I see it makes the same mistake.

**Reply by James-Firth** [reply]
@willfarrell I'd appreciate an example! Is this code from v6?

I don't see `openpgp.stream` nor is example in the online docs (I get a `verificationResults.data` is not iterable issues in the `for await...` section of the Uint8Array sign/verify example)

**Reply by larabr** [reply]
@James-Firth the code posted here is related to a pre-release of v5, hence the code no longer works as-is. `openpgp.stream` is no longer  available in v5, you can instead import [`@openpgp/web-stream-tools`](https://github.com/openpgpjs/web-stream-tools). Some options in `sign` and `verify` have been changed to (e.g. `publicKeys` renamed to `verificationKeys`, and `privateKeys` renamed to `signingKeys`).
